# Agenda:

1. What is Machine Learning?
2. What are the different types of Machine Learning?
3. Supervised Machine Learning
4. UnSupervised Machine Learning
5. Reinforcement Machine Learning

# 1. What is Machine Learning?

- In simple words, ML can be defined as teaching the machine to learn from the historical data.
- Its a study of teaching a computer program/an algorithm how to progressively improve upon a set that the is given to it.


## Traditional way of programming(period before 2012 or 2010):
- Software Developers write the code to build the applications.

## Machine Learning(period after 2010 or so) :
- Teach the machine to write the code in order to find the pattern and develop an application.
- You supply the data to a ML algorithm, ML algorithm itself will learn from the data and develop conclusions out of the data

## Types of ML Algorithms:
1. Supervised Learning
2. UnSupervised Learning
3. Reinforcement Learning  
- And also there is some thing called as semi supervised learning

- Machine Learning Algorithms are the "engines of Machine Learning", meaning it the algorithms that turn a dataset into a Machine Leanring Model

- Which type of ML algorithms are the best?
- It completely depends on the problem that we are currently solving, the computing resources available and the nature of the data

# Supervised Learning:

- It is widely used when you have both input and output available inside your data

### Ways to apply Supervised ML on the dataset:

- We divide the data into 2 parts:
    1. Training data
    2. Test data
- How to divide into train and test data?
- The idea the training data should more and test data should be less(70% train and 30% test)
(80% train and 20% test)
- Training data is where the machine Learning algorithms are applied to learn the patterns from the data.
- Test data is used to check how good the ML algorithm was able to learn the patterns.

In [1]:
import seaborn as sns
sns.load_dataset('iris')

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

In [6]:
150 * 0.8, 150 * 0.2

(120.0, 30.0)

In [7]:
150 * 0.7, 150 * 0.3

(105.0, 45.0)

- As an example, if we try to divide the iris data into 80% train and 20% test, the train data will have 120 rows and the test data will have 30 rows.
- In same way if we try to divide the iris data into 70% train and 30% test, the train data will have 105 rows and test data will have 45 rows.

# Applications of Supervised ML:
- Sales prediction
- Salary prediction
- Diseases prediction
- Win or lose prediction

# Types of Supervised ML:

1. Regression - prediction of a continuous value based on the input variables
- eg: prediction of tip from the tips dataset, prediction of salary, prediction of runs that India will score, etc

2. Classification - predicting a class or category based on the input variables

- eg: prediction of iris flower, predicting whether a person has a disease or not, predicting whether India wins or looses the match, etc.

In [8]:
sns.load_dataset('tips')

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

# Unsupervised Machine Learning:

- The data will not have any output value.
- The main idea of Unsupervised Machine Learning Algorithms is grouping/Clustering/Segmentation
- YouTube Vedio Recommendations, Netflix Movie Recommendation, Amazon Product Recommendations, DMART product Recommendation, etc

In [ ]:
stud_id stud_name stud_gender age height weight marks
1        A         M           15  5.8     87    85
2        B         F           15  5.8     70    95
3        C         F           15  5.8     65    60

>80 Section A
> 60 and <80 Section B
>40 and < 60 Section C
< 40 Section D

# Reinforcement Learning

- It works on the principle or punish and reward.
- Most popularly used in robotics
- Even in online games

# The focus will be on Supervised ML algorithms
1. Regression - Linear Regression, Decision Tree Regression, etc
2. Classification - Logistic REgression, Decision Tree Classificaton, etc

# Unsupervised : 
1. K - means algorithm
2. DBSCAN
3. Hierarchial Clustering
4. Mean Shift

# Terminology of ML

- Columns : Variable, Features
- Rows : Records, observation, example

# Terminology of Supervised ML:

- Input Variables - Independent Variables or Independent Features
- Ouput Variable - Dependent Variable, Target, output

- eg: In iris dataset, sepal_length, sepal_widht, petal length and petal_width are all Independent variables. species is the output or target
- eg: In tips datasets, tip variable is the target or the output and all other variable are independent

## Note: For any dataset, There must be only 1 column which we call as target. We cannot have multiple columns as target.

# Solving the Machine Learning Problems:

1. Define the problem for the dataset
2. Loading the data
3. Data Cleaning, Data Wrangling, Data Visualizaton, Data Analysis
4. Data Preprocessing
5. Seperate the independent variables and the target
6. Divide the data into train data and test data
7. Apply ML algorithm on the train data
8. Evaluate the model on the test data.
9. Perform Accuracy check

eg:
1. Defining a problem for iris data: Given sepal_length, sepal_width, petal_length and petal_width predict the species of the iris flower.
2. Load the data